processor/intracesampler/processor.go

```diff
@@ -0,0 +1,224 @@
+	children map[pcommon.SpanID][]pcommon.SpanID
+
+	// spans that have no parent or their parent is not in the trace
+	roots []pcommon.SpanID
+}
+
+// this map enable us to find the parent span of a span in O(1), and all the children of a givin span
+// it also generates a span object that contains the resource and scope all at once
+// this is useful for the sampler to be able to make decisions on spans
+func spansToTraceTree(td ptrace.Traces) TraceTreeData {
+	fullSpans := make(map[pcommon.SpanID]FullSpan)
+	spanChildren := make(map[pcommon.SpanID][]pcommon.SpanID)
+
+	rss := td.ResourceSpans()
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+		resource := rs.Resource()
+		scopespans := rs.ScopeSpans()
+		for j := 0; j < scopespans.Len(); j++ {
+			ss := scopespans.At(j)
+			spans := ss.Spans()
+			scope := ss.Scope()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+
+				fullspan := FullSpan{
+					resource: resource,
+					scope:    scope,
+					span:     span,
+				}
+
+				fullSpans[span.SpanID()] = fullspan
+				spanChildren[span.ParentSpanID()] = append(spanChildren[span.ParentSpanID()], span.SpanID())
+			}
+		}
+	}
+
+	// find roots
+	roots := make([]pcommon.SpanID, 0)
+	for _, fullspan := range fullSpans {
+		parentSpanID := fullspan.span.ParentSpanID()
+		if _, ok := fullSpans[parentSpanID]; !ok {
+			currentSpanID := fullspan.span.SpanID()
+			roots = append(roots, currentSpanID)
+		}
+	}
+
+	traceTreeData := TraceTreeData{
+		fullSpans: fullSpans,
+		children:  spanChildren,
+		roots:     roots,
+	}
+
+	return traceTreeData
+}
+
+// check if all spans in td are from the span trace id.
+// this indicates that the processor is run after another processor
+// that emits completed traces after timeout
+// if a single trace id is found, it is returend, otherwise nil is returned
+func getSingleTraceID(td ptrace.Traces) *pcommon.TraceID {
+	var traceID *pcommon.TraceID
+	rss := td.ResourceSpans()
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+		scopespans := rs.ScopeSpans()
+		for j := 0; j < scopespans.Len(); j++ {
+			ss := scopespans.At(j)
+			spans := ss.Spans()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+				currentTraceID := span.TraceID()
+				if traceID == nil {
+					traceID = &currentTraceID
+				} else if currentTraceID != *traceID {
+					return nil
+				}
+			}
+		}
+	}
+
+	// will be nil it the batch is empty
+	return traceID
+}
+
+func (its *inTraceSamplerProcessor) getScopeBranchesToUnsampleRec(traceTreeData TraceTreeData, currentSpanID pcommon.SpanID, unsampledScopes map[pcommon.SpanID]bool) bool {
+	currentFullSpan := traceTreeData.fullSpans[currentSpanID]
+	currentScopeName := currentFullSpan.scope.Name()
+
+	// currrent span should be unsampled if it's in the unsampledScopes map
+	// and all its children are also unsampled.
+	currentUnsampled := slices.Contains(its.config.ScopeLeaves, currentScopeName)
```

Comment:
I feel like you would be better of using map like a set instead of `slice.Contains`.

You can use:
```golang
map[pcommon.SpanID]struct{}
```

This will allocate no memory for the empty struct and will be more optimal than `slices.Contains`

Reply:
Fixed. I hope I implemented it in a go idiomatic manner (I am new to golang).

If you have the time to take another look that would be amazing